lib/instrumentation/undici.js

```diff
@@ -0,0 +1,291 @@
+/*
+ * Copyright 2020 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+const recordExternal = require('../metrics/recorders/http_external')
+const hashes = require('../util/hashes')
+const logger = require('../logger').child({ component: 'undici' })
+const NAMES = require('../metrics/names')
+const SHIM_SYMBOLS = require('../shim/constants').SYMBOLS
+
+const NEWRELIC_ID_HEADER = 'x-newrelic-id'
+const NEWRELIC_TRANSACTION_HEADER = 'x-newrelic-transaction'
+const NEWRELIC_SYNTHETICS_HEADER = 'x-newrelic-synthetics'
+const diagnosticsChannel = require('diagnostics_channel')
+const SEGMENT = Symbol('__NR_segment')
+const PARENT_SEGMENT = Symbol('__NR_parent_segment')
+const { TLSSocket } = require('tls')
+
+/**
+ * Subscribes to undici diagnostic channel events
+ *  `undici:request:create` - happens right before request is made
+ *  `undici:request:headers` - happens when response headers are returned from server
+ *  `undici:request:trailer` - happens right before response ends
+ *  `undici:request:error` - happens when request errors
+ *
+ * @param {Agent} agent
+ */
+module.exports = function addUndiciChannels(agent, undici, modName, shim) {
+  diagnosticsChannel.channel('undici:request:create').subscribe(({ request }) => {
+    const parent = agent.tracer.getSegment()
+    request[PARENT_SEGMENT] = parent
+    if (parent && parent.opaque) {
+      logger.trace(
+        'Not capturing data for outbound request (%s) because parent segment opaque (%s)',
+        request.path,
+        parent.name
+      )
+
+      return
+    }
+
+    const transaction = parent.transaction
+    const outboundHeaders = Object.create(null)
+    if (agent.config.encoding_key && transaction.syntheticsHeader) {
+      outboundHeaders[NEWRELIC_SYNTHETICS_HEADER] = transaction.syntheticsHeader
+    }
+
+    // TODO: abstract header logic shared with TransactionShim#insertCATRequestHeaders
+    if (agent.config.distributed_tracing.enabled) {
+      // TODO: show can a request header be a symbol??
+      if (request.headers && request.headers[SHIM_SYMBOLS.DISABLE_DT]) {
```

Comment:
yea i noticed that. Honestly this is copied code from our http-outbound and I don't see how a request header could ever be a symbol when serialized